mqtt_io/server.py

```diff
@@ -1191,16 +1169,48 @@ async def run() -> None:
                 try:
                     await asyncio.gather(*self.critical_tasks)
                 except asyncio.CancelledError:
-                    pass
+                    break
                 except Exception:  # pylint: disable=broad-except
                     _LOG.exception("Exception in critical task:")
             except asyncio.CancelledError:
-                pass
+                break
+            except MQTTException:
+                if counter > 0:
+                    counter -= 1
+                _LOG.exception('Connection to MQTT-Broker failed')
+                if counter == 0:
+                    break
+                await asyncio.sleep(self.config['mqtt'].get('reconnect_delay'))
```

Comment:
Might be best to just set a flag here saying that we'll reconnect or not, so that the `finally` block can run as soon as we're disconnected, clearing the `self.running` event. We should also clear the `self.mqtt_connected` event in the `finally` block too.

Then once we're out of that block, we can sleep if we're reconnecting and then loop back around. (We could also use that flag as the `while` predicate too, so:

```python
reconnect = True
while reconnect:
    try:
        await self._connect_mqtt()
        ...
    except MQTTException:
        _LOG.exception('Connection to MQTT broker failed')
        reconnect = counter > 0
        counter -= 1
    finally:
        self.running.clear()
        self.mqtt_connected.clear()
    if reconnect:
        await asyncio.sleep(...)
```

Something like that, anyway!

Reply:
Also, since we'll have probably popped out of `await asyncio.gather(*self.critical_tasks)` with an MQTTException, we'll need to cancel the rest of the Tasks in `self.critical_tasks`, then gather those before we can loop back around and create new ones. I'd maybe do this in the `finally` block, since we need to do that whether we're reconnecting or not.

I need to spend some more time on keeping track of Tasks, having recently been using Trio and learned a lot more about how async Task management should work, but this should do for now.

Reply:
Ok, I've changed that

Reply:
I didn't see your 2. message earlier, I also changed that now.

And I think the Task management should be reworked a little bit, too. Some of the tasks could run independently of the reconnect-loop.